packages/hebao_v3/contracts/base/libwallet/ERC20Lib.sol

```diff
@@ -0,0 +1,315 @@
+        bytes32          domainSeperator,
+        Request calldata request,
+        address          token,
+        address          to,
+        uint             amount,
+        uint             value,
+        bytes   calldata data
+        )
+        external
+        returns (bytes memory returnData)
+    {
+        bytes memory encoded = abi.encode(
+            APPROVE_THEN_CALL_CONTRACT_TYPEHASH,
+            request.wallet,
+            request.validUntil,
+            token,
+            to,
+            amount,
+            value,
+            keccak256(data)
+        );
+
+        wallet.verifyRequest(
+            domainSeperator,
+            SigRequirement.MAJORITY_OWNER_REQUIRED,
+            request,
+            encoded
+        );
+
+        _approveInternal(token, to, amount);
+        return _callContractInternal(to, value, data, PriceOracle(0));
+    }
+
+    function transfer(
+        address token,
+        address to,
+        uint    amount
+        )
+        public
+    {
+        if (token == address(0)) {
+            to.sendETHAndVerify(amount, gasleft());
+        } else {
+            ERC20(token).safeTransfer(to, amount);
+        }
+    }
+
+    // --- Internal functions ---
+
+    function _transferWithEvent(
+        address token,
+        address to,
+        uint    amount,
+        bytes   calldata logdata
+        )
+        private
+    {
+        transfer(token, to, amount);
+        emit Transfered(token, to, amount, logdata);
+    }
+
+    function _approveInternal(
+        address token,
+        address spender,
+        uint    amount
+        )
+        private
+        returns (uint additionalAllowance)
+    {
+        // Current allowance
+        uint allowance = ERC20(token).allowance(address(this), spender);
+
+        if (amount != allowance) {
+            // First reset the approved amount if needed
+            if (allowance > 0) {
+                ERC20(token).safeApprove(spender, 0);
+            }
+            // Now approve the requested amount
+            ERC20(token).safeApprove(spender, amount);
+        }
+
+        // If we increased the allowance, calculate by how much
+        if (amount > allowance) {
+            additionalAllowance = amount.sub(allowance);
+        }
+        emit Approved(token, spender, amount);
+    }
+
+    function _callContractInternal(
+        address              to,
+        uint                 value,
+        bytes       calldata txData,
+        PriceOracle          priceOracle
+        )
+        private
+        returns (bytes memory returnData)
+    {
+        if (priceOracle != PriceOracle(0)) {
+            // Disallow general calls to token contracts (for tokens that have price data
+            // so the quota is actually used).
+            require(priceOracle.tokenValue(to, 1e18) == 0, "CALL_DISALLOWED");
+        }
+
+        bool success;
+        (success, returnData) = to.call{value: value}(txData);
```

Comment:
maybe we should check `to != address(this)`?

Reply:
Also a case where it should be fine in theory, but can't see any benefit in practice so will disallow.